Add linear friction functor material

snschune opened the pull request:
closes #24702 

## Reason

It should be easy to define a quadratic loss in PINSFV.

## Design

A functormaterial that does that.

## Impact

see #24702

Commit:
Add linear friction functor material (#24702)

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1592166/) on e70f199 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/24703/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/24703/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 23ac2e6b1e6eca40969c8f4bacad2a8d5112b4a1`

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1592247/) on 31989a8 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>23ac2e</th>
      <th colspan=4>#24703 31989a</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.41%</td>
      <td>85.40%</td>
      <td>-0.00%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>88937</td>
      <td>88934</td>
      <td>-3</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15196</td>
      <td>15199</td>
      <td>+3</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24703/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24703/coverage/framework)


## Modules coverage

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>23ac2e</th>
      <th colspan=4>#24703 31989a</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>83.87%</td>
      <td>83.91%</td>
      <td>+0.04%</td>
      <td>95.24%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>10836</td>
      <td>10876</td>
      <td>+40</td>
      <td>40</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>2084</td>
      <td>2086</td>
      <td>+2</td>
      <td>2</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24703/coverage_diff/navier_stokes)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24703/coverage/navier_stokes)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/24703/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/24703/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/24703/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/24703/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/24703/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/24703/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/24703/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/24703/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/24703/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/24703/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/24703/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/24703/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/24703/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/24703/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/24703/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/24703/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/24703/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/24703/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/24703/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/24703/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/24703/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/24703/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/24703/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/24703/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/24703/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/24703/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/24703/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/24703/coverage/xfem)

</details>



This comment will be updated on new commits.

GiudGiud commented on modules/navier_stokes/src/functormaterials/LinearFrictionFactorFunctorMaterial.C:
```diff
@@ -0,0 +1,84 @@
+  params.addParam<MooseFunctorName>(
+      NS::superficial_velocity_z, 0, "The z component of the fluid superficial velocity variable.");
+  params.addParam<RealVectorValue>(
+      "A", RealVectorValue(1, 1, 1), "Coefficient of the A * f(t) term");
+  params.addParam<RealVectorValue>(
+      "B", RealVectorValue(1, 1, 1), "Coefficient of the B * g(t) * |v_I| term");
```
no defaults. or defaults of 0

too risky if a user wants linear and we re adding them a quadratic by default

GiudGiud commented on modules/navier_stokes/src/functormaterials/LinearFrictionFactorFunctorMaterial.C:
```diff
@@ -0,0 +1,84 @@
+{
+  InputParameters params = FunctorMaterial::validParams();
+  params.addClassDescription("Material class used to compute a friction factor of the form A * "
+                             "f(t) + B * g(t) * |v_I| with A, B vector constants, f(t) and g(t) "
```
```suggestion
                             "f(r, t) + B * g(r, t) * |v_I| with A, B vector constants, f(r, t) and g(r, t) "
```

GiudGiud commented on modules/navier_stokes/src/functormaterials/LinearFrictionFactorFunctorMaterial.C:
```diff
@@ -0,0 +1,84 @@
+  InputParameters params = FunctorMaterial::validParams();
+  params.addClassDescription("Material class used to compute a friction factor of the form A * "
+                             "f(t) + B * g(t) * |v_I| with A, B vector constants, f(t) and g(t) "
+                             "functions of time, and |v_I| the interstitial speed");
```
```suggestion
                             "functors of space and time, and |v_I| the interstitial speed");
```

GiudGiud commented on modules/navier_stokes/test/tests/finite_volume/pins/channel-flow/tests:
```diff
@@ -187,4 +187,13 @@
+    type = 'Exodiff'
+    input = linear_friction.i
+    exodiff = linear_friction_out.e
+    issues = '#24703'
```
```suggestion
    issues = '#24702'
```

GiudGiud commented on modules/navier_stokes/doc/content/source/functormaterials/LinearFrictionFactorFunctorMaterial.md:
```diff
@@ -0,0 +1,24 @@
+!equation
+-\epsilon \nabla p = W \rho \vec{v}_I
+
+in the porous flow equations where all symbols have their common meaning. The
```
```suggestion
in the porous flow equations where all symbols have their common meaning (see [porous media equations](navier_stokes/pinsfv.md)) and we do not write all the other momentum equation terms. The
```

lindsayad commented on modules/navier_stokes/src/functormaterials/LinearFrictionFactorFunctorMaterial.C:
```diff
@@ -0,0 +1,84 @@
+      {
+        // Compute speed
+        // see PINSFVSpeedFunctorMaterial.C for explanation
+        ADReal speed;
+        if ((MooseUtils::absoluteFuzzyEqual(_superficial_vel_x(r, t), 0)) &&
+            (MooseUtils::absoluteFuzzyEqual(_superficial_vel_y(r, t), 0)) &&
+            (MooseUtils::absoluteFuzzyEqual(_superficial_vel_z(r, t), 0)))
+          speed = 1e-42;
+
+        speed = ADRealVectorValue(
+                    _superficial_vel_x(r, t), _superficial_vel_y(r, t), _superficial_vel_z(r, t))
+                    .norm() /
+                _eps(r, t);
```
I would save off a local velocity variable and then call `NS::computeSpeed`
```suggestion
        const ADRealVectorValue superficial_vel(_superficial_vel_x(r, t),
                                                _superficial_vel_y(r, t),
                                                _superficial_vel_z(r, t));
        const auto speed = NS::computeSpeed(superficial_vel) / _eps(r, t);
```

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1596950/) on 4f1fe35 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/24703/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/24703/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 20e735bcdd25ae44efc67790e23091d2eed7ff00`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1597381/) on 4b6b6ba wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/24703/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/24703/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 20e735bcdd25ae44efc67790e23091d2eed7ff00`

moosebuild commented:
Job [OpenMPI](https://civet.inl.gov/job/1597565/) on 31989a8 : invalidated by @lindsayad

snschune committed:
Address reviews (#24702)

- Remove unformatted description from .md file
- Add like to PINSFV equations
- Use more `const`
- Use NS utility for speed computation
- Remove perturbation to velocity initial condition (required regold)
- Correct issue reference

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1616385/) on b2381e4 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/24703/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/24703/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 0e3d4f264182cdd0d2b4392368092302fad62522`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1616385/) on b2381e4 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/24703/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/24703/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 0e3d4f264182cdd0d2b4392368092302fad62522`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1616429/) on f1bea20 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/24703/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/24703/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 0e3d4f264182cdd0d2b4392368092302fad62522`

lindsayad commented on modules/navier_stokes/doc/content/source/functormaterials/LinearFrictionFactorFunctorMaterial.md:
```diff
@@ -0,0 +1,24 @@
+# LinearFrictionFactorFunctorMaterial
+
+!syntax description /FunctorMaterials/LinearFrictionFactorFunctorMaterial
+
```
I would probably just remove these lines as it introduces the equation without formatting whereas you have it with formatting later in this file

lindsayad commented on modules/navier_stokes/include/functormaterials/LinearFrictionFactorFunctorMaterial.h:
```diff
@@ -0,0 +1,46 @@
+  MooseFunctorName _functor_name;
+
+  ///@{ A, B, f(t), g(t)
+  RealVectorValue _A;
+  RealVectorValue _B;
```
```suggestion
  const RealVectorValue _A;
  const RealVectorValue _B;
```

lindsayad commented on modules/navier_stokes/test/tests/finite_volume/pins/channel-flow/linear_friction.i:
```diff
@@ -0,0 +1,101 @@
+    dynamic_viscosity = 'mu'
+    porosity = 'porosity'
+
+    initial_velocity = '1 1e-6 0'
```
We probably don't need these small perturbations from a 0 initial condition. We may have needed them at some point, but I hope we don't still

lindsayad reviewed (commented):
Just small comments

lindsayad commented:
@snschune I will fix the exodiff and squash some of the commit history

GiudGiud reviewed (approved):
looks good!
